### PR TITLE
Added project permissions and GitHub push integration

### PR DESCRIPTION
**Pull request by barankyle**
## Summary

project-permission is a structure that gives a user 'owner' or 'user' access to a project.
Regular project-users will not be able to grant or remove project-permissions to/from other
users, and might not be able to delete the project.

People with access to a project will be able to edit its files, but will be prevented from
editing files in projects they do not have access to.

## References

#5613

## Checklist
- [x] If this PR is still a WIP, convert to a draft 
- [x] When this PR is ready, mark it as "Ready for review"
- [x] Changes have been manually QA'd
- [x] Changes reviewed by at least 2 approved reviewers

## QA Steps

_List any additional steps required to QA the changes of this PR, as well as any supplemental images or videos._

### CODE REVIEW

**Comment by barankyle** [comment]
Instructions for testing:

Make sure that the local GH app credentials are installed. I will send the environment variables and .pem via DM. To install the .pem, go to /admin/settings, select Server, then copy/paste the entirety of the .pem contents into the field "Github Private Key". Make sure to save.

On /editor, opening the project menu via the gear will have more options. You can view the permissions from 'Project Permissions'; if you are an owner of a project, you can add other users via invite code, change their permissions from 'user' to 'owner' (and vice-versa), and delete project permissions. Any project that you have access to, you should be able to link it to a GH repo if it does not have a repositoryPath, or change/delete the link if it does. You can also trigger a download from GH. If you have a project-permission for a project, or if you are logged in with GH credentials and have write access to the repo that the project is linked to, you should be able to commit + push it to the GH repo.

On /admin/projects, all of the above functionality should be basically identical, just accessed via new columns in the table that open drawers.

In the editor itself, you should only have access to project files that you are authorized for.

File: packages/client-core/src/common/services/ProjectService.ts [hunk 1]
```diff
@@ -73,17 +77,73 @@ export const ProjectService = {
...
+      throw err
+    }
   }
 }
 // TODO
 // client.service('project-build').on('patched', (params) => {
 //   store.dispatch(ProjectAction.buildProgress(params.message))
 // })
 
+API.instance.client.service('project').on('patched', (params) => {
```
Comment (HexaField) [inline packages/client-core/src/common/services/ProjectService.ts:143]: this needs to be moved to the service in a `useAPIListeners()` as elsewhere in the codebase

**Review by HexaField — CHANGES_REQUESTED** [review]

File: packages/client-core/src/common/services/ProjectService.ts (hunk 1 above)
Comment (barankyle) [reply packages/client-core/src/common/services/ProjectService.ts]: Check now, I think I did it right.

Comment (HexaField) [reply packages/client-core/src/common/services/ProjectService.ts]: almost, need to pull the listener out and clean it up on the useEffect unmount return function

**Comment by hanzlamateen** [comment]
The user access dialog is appearing as empty any idea? I do followed the above "Instructions for testing"

image

**Comment by hanzlamateen** [comment]
Also the push button is disabled on my end. How can I enable it?

**Comment by hanzlamateen** [comment]
> I made a few updates to address your issues.
> 
> Admins can now make permissions on a project even if they are not an owner - you should see a field to grant permission to a user by inviteCode. You can grant it to yourself, though you'll need an inviteCode (i.e. be a non-guest, or not the initial guest user who's made an admin). You might have to manually add an inviteCode to the initial guest/admin.
> 
> The push button only becomes active if a) there's a value for repositoryPath and b) the user has a project-permission for the project, or has a GitHub identity-provider and has write access to the repo for that project. Since you can now make project-permissions, you should be able to see it light up now.
> 
> If you want to test pushing to something other than the actual repo where a project is located (which I'd recommend so we don't litter our main repos with unnecessary commits), you'll need to install this app into the org/your personal account where you're testing it: https://github.com/apps/theoverlay-local

I installed the app on my personal repo and then forked project (https://github.com/hanzlamateen/XREngine-development-test-suite). Added this project from admin panel.

Made some changes in Readme and xrengine config of that project and tried to use push button from admin panel. No commit is pushed to repo andthe page is automatically refreshed when I confirm from dialog.

**Comment by hanzlamateen** [comment]
It seems like service exposes property `project_permission` whereas client is using `project_permissions` there is an additional 's' being used in client. Is it desired?

**Comment by hanzlamateen** [comment]
/home/hanzlamateen/Videos/simplescreenrecorder-2022-06-30_08.35.00.mp4

**Comment by hanzlamateen** [comment]
https://user-images.githubusercontent.com/10975502/176587510-9cd5c118-4fd9-410f-bb70-cbd1db936f5f.mp4

**Comment by hanzlamateen** [comment]
image

Nothing happens when I click on project permissions item in the menu

**Comment by barankyle** [comment]
> 

sequelize will automatically pluralize related fields when they're populated like this, so each project will have a field "project_permission**s**" on it when it's sent to the client.

**Comment by barankyle** [comment]
> Made some changes in Readme and xrengine config of that project and tried to use push button from admin panel. No commit is pushed to repo and the page is automatically refreshed when I confirm from dialog.

The refresh comes from vite in dev mode watching the projects, among other things, and restarting when it detects changes there. If 

As for that commit not being pushed, I think it's one of two things: Either you didn't set the .pem file in admin/settings -> Server -> GitHub Private Key, or this project has files that are too large. GitHub has a soft cap of 50MB on files, and while it seems to accept files between 50-100MB when pushed via the cli, the API is throwing an error when our backend tries to create a blob of a file that large. Try again without any files > 50MB and it should work.

**Comment by barankyle** [comment]
> Nothing happens when I click on project permissions item in the menu

This was a mistake, those options are no longer shown on non-installed projects.

**Comment by hanzlamateen** [comment]
> simplescreenrecorder-2022-06-30_08.35.00.mp4 
> Inconsistency in user permission. Please watch and listen above video

@barankyle what about this one?

**Review by hanzlamateen — APPROVED** [review]

**Pull request merged**